### PR TITLE
add ActivityEvent, keepInfo.activityLog

### PR DESCRIPTION
**Pull request by camhashemi**
@andrewconner @ryanpbrewster @leogrim 

this is just a start so that clients can start working with it, adding initial keep events next

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala [hunk 1]
```diff
@@ -338,6 +342,8 @@ class KeepDecoratorImpl @Inject() (
       }
     }
   }
+
+  def generateActivityLog(discussion: Discussion): Seq[ActivityEvent] = discussion.messages.map { msg => ActivityEvent.fromComment(msg) }
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala]: this is a start so that clients can start working with and giving feedback on the new format, next step for me is to add the initial keep events ("Cam added this keep to LibraryX" with note)

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+sealed abstract class ActivityKind(val value: String)
+object ActivityKind extends Enumerator[ActivityKind] {
+  case object Initial extends ActivityKind("initial")
+  case object Comment extends ActivityKind("comment")
+  case object AddedParticipants extends ActivityKind("added_participants")
+  case object AddedLibrary extends ActivityKind("added_library")
+  case object EditedTitle extends ActivityKind("edited_title")
+
+  def apply(str: String) = _all.find(_.value == str).get
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala]: I think you want to define `val all = _all` somewhere, `_all` is a macro so it may be expensive to call it every time.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+object ActivitySource extends Enumerator[ActivitySource] {
+  case object Slack extends ActivitySource("Slack")
+  case object Twitter extends ActivitySource("Twitter")
+  case object iOS extends ActivitySource("iOS")
+  case object Android extends ActivitySource("Android")
+  case object Chrome extends ActivitySource("Chrome")
+  case object Firefox extends ActivitySource("Firefox")
+  case object Safari extends ActivitySource("Safari")
+
+  def apply(str: String) = _all.find(_.value == str).get
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala]: same deal, `val all = _all` or something

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  object TextElement {
+    val kind = "text"
+    implicit val writes = new Writes[TextElement] { def writes(t: TextElement) = Json.obj("text" -> t.text, "kind" -> kind) }
+  }
+
+  case class UserElement(
+    id: ExternalId[User],
+    name: String,
+    image: String,
+    path: String) extends ActivityElement
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala]: `path` can be a `Path`?

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 3 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala]: :+1: thanks

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  implicit def writes = Writes[ActivityElement] {
+    case txt: TextElement => TextElement.writes.writes(txt)
+    case usr: UserElement => UserElement.writes.writes(usr)
+    case nur: NonUserElement => NonUserElement.writes.writes(nur)
+    case lib: LibraryElement => LibraryElement.writes.writes(lib)
+    case org: OrganizationElement => OrganizationElement.writes.writes(org)
+  }
+}
+
+sealed case class ActivityEvent(
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala]: Probably `final`, not `sealed`, here, unless you are subclassing it.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      image = msg.sentBy.right.toOption.map(_.pictureName).getOrElse("0.jpg"),
+      header = Seq(msgAuthor, ActivityElement.TextElement("commented on this page")),
+      body = Seq(ActivityElement.TextElement(msg.text)),
+      bodyText = msg.text,
+      timestamp = msg.sentAt,
+      source = None // todo(cam): add eliza.message.source (column) to Message
+    )
+  }
+
+  implicit def writes: Writes[ActivityEvent] = (
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala]: can be a `val`

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  bodyText: String, // message and keep.note content fallback
+  timestamp: DateTime,
+  source: Option[ActivitySource])
+object ActivityEvent {
+  def fromComment(msg: Message): ActivityEvent = {
+    val msgAuthor: ActivityElement = msg.sentBy.fold(ActivityElement.fromNonUser, ActivityElement.fromUser)
+    ActivityEvent(
+      ActivityKind.Comment,
+      image = msg.sentBy.right.toOption.map(_.pictureName).getOrElse("0.jpg"),
+      header = Seq(msgAuthor, ActivityElement.TextElement("commented on this page")),
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: May be nice, eventually, to have something like @leogrim's `DescriptionElements` with their various constructors.

File: kifi-backend/modules/common/test/com/keepit/common/util/ActivityElementsTest.scala [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import org.specs2.mutable.Specification
+import play.api.libs.json.Json
+
+class ActivityElementsTest extends Specification {
+  "ActivityElements" should {
+    "format properly" in {
+      val text = ActivityElement.TextElement("hey")
+      val user = ActivityElement.UserElement(ExternalId[User](), "Jimmy", "0.jpg", "/jimmy")
+      TestHelper.deepCompare(Json.toJson(text)(ActivityElement.writes), Json.obj("text" -> text.text, "kind" -> ActivityElement.TextElement.kind)) === None
+      TestHelper.deepCompare(Json.toJson(user), Json.obj("id" -> user.id, "name" -> user.name, "image" -> user.image, "path" -> user.path, "kind" -> ActivityElement.UserElement.kind))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/test/com/keepit/common/util/ActivityElementsTest.scala]: `must beNone` here

File: kifi-backend/modules/common/app/com/keepit/model/UserExperiment.scala [hunk 9]
```diff
@@ -82,6 +82,7 @@ object UserExperimentType {
   val DISCUSSION_FEED_FILTERS = UserExperimentType("discussion_feed_filters")
   val KEEP_PAGE_RHR = UserExperimentType("keep_page_rhr")
   val PRIVATE_SYNC = UserExperimentType("private_sync")
+  val ACTIVITY_LOG = UserExperimentType("activity_log")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/UserExperiment.scala:85]: Is this experiment checked anywhere?

**Comment by ryanpbrewster** [comment]
lgtm, nice work. Do you know if we can unify some of these `X_Source` entities (`KeepSource`, `MessageSource`, etc.)?

File: kifi-backend/modules/common/app/com/keepit/model/UserExperiment.scala (hunk 9 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/UserExperiment.scala:85]: clients will use it to determine how keeps should be rendered (using keep.discussion or keep.activityLog)

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/UserExperiment.scala:85]: Ah right, I always forget that the front-end uses the experiments too

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 7 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: yeah definitely going to be doing that as the `ActivityEvent` generation picks up.

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: Yep I'm not sure what requires this new type? Could you explain?

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+
+sealed abstract class ActivityKind(val value: String)
+object ActivityKind extends Enumerator[ActivityKind] {
+  case object Initial extends ActivityKind("initial")
+  case object Comment extends ActivityKind("comment")
+  case object AddedParticipants extends ActivityKind("added_participants")
+  case object AddedLibrary extends ActivityKind("added_library")
+  case object EditedTitle extends ActivityKind("edited_title")
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:20]: So how is this related to / different from `SystemMessageData`?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala [hunk 11]
```diff
@@ -338,6 +342,8 @@ class KeepDecoratorImpl @Inject() (
       }
     }
   }
+
+  def generateActivityLog(discussion: Discussion): ActivityLog = ActivityLog(discussion.messages.map { msg => ActivityEvent.fromComment(msg) })
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala:346]: Ah ok cool!

**Comment by leogrim** [comment]
@camhashemi I like how you started from the API as to define a clear contract!
I'm not so sure about `ActivityElement`. There are two approaches we can take here:
- fully typed case-class-style events, and the front-end knows to handle each of them, simply ignores the kinds it doesn't know.
- fully generic (like our payment `ActivityLog`) in which case we should reuse `DescriptionElements` to send down markdown-style content, not typed-entities.

`ActivityElement` feels in between, introducing new types that the front-end still needs to know how to handle case-by-case, while passing them as generic lists. 

Curious to hear more from @ryanpbrewster @andrewconner

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+sealed abstract class ActivitySource(val value: String)
+object ActivitySource extends Enumerator[ActivitySource] {
+  case object Slack extends ActivitySource("Slack")
+  case object Twitter extends ActivitySource("Twitter")
+  case object iOS extends ActivitySource("iOS")
+  case object Android extends ActivitySource("Android")
+  case object Chrome extends ActivitySource("Chrome")
+  case object Firefox extends ActivitySource("Firefox")
+  case object Safari extends ActivitySource("Safari")
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:36]: Nice, I like that you're treating `Slack` and `Twitter` just like clients

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    case nur: NonUserElement => NonUserElement.writes.writes(nur)
+    case lib: LibraryElement => LibraryElement.writes.writes(lib)
+    case org: OrganizationElement => OrganizationElement.writes.writes(org)
+  }
+}
+
+case class ActivityEvent(
+  kind: ActivityKind,
+  image: String,
+  header: Seq[ActivityElement], // e.g. "Cam kept this in LibraryX"
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:126]: This looks generic -from the field names- but in practice the front-end still needs to understand the semantics of the event to render it.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 7 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: `DescriptionElements` flattens objects into text on the server side, whereas here we're passing json objects down to clients to render how they want. Minus the text elements and overall structure, I don't think there is much re-usable between the two.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 10 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:20]: My big question right now is whether we should be adding to `SystemMessageData` or going with `KeepToUser` and `KeepToLibrary` as a reference for events. In general it seems like we want to stay away from `eliza` since she doesn't know much about permissions and model info, but as you're suggesting `SystemMessageData` can be built upon to introduce new events.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 7 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: `DescriptionElements` actually does go down to clients as a `JsArray`.

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]:
```
Json.prettyPrint(DescriptionElements.flatWrites.writes(DescriptionElements(
    "one fish",
    "two fish" --> LinkElement("www.google.com"),
    "red fish blue fish" --> LinkElement("www.youngrepublicans.com")
                         --> Hover("cool", "hover board" --> LinkElement("www.amazon.com"))
))) === [ {
  "text" : "one fish "
}, {
  "text" : "two fish",
  "url" : "www.google.com"
}, {
  "text" : " "
}, {
  "text" : "red fish blue fish",
  "url" : "www.youngrepublicans.com",
  "hover" : [ {
    "text" : "cool "
  }, {
    "text" : "hover board",
    "url" : "www.amazon.com"
  } ]
} ]
```

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 13 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:126]: Why do they need to know the semantics?

Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:126]: Which semantics are you referring to? I can think of hashtags in notes and look heres in messages, but those behaviors can be specified in `LookHere` and `HashTag` `ActivityElements`.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 7 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: Ah, I haven't seen that. Where is this used? I think we could go down this route, but that there are so many helper methods and assumptions in `DescriptionElements` that don't apply to the markup that @andrewconner and @DerekBurch discussed with me. I would be either making a different markup than discussed or making `DescriptionElements` less usable.

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:136]: It's used in the billing ActivityLog and for CreditReward

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 13 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:126]: You're right, front-end doesn't need to know the semantics. But needs to know about each type of `ActivityElement`

Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:126]: They need to anyways, e.g. look heres, hashtags, org and library chips, and any inline features to come. There's specific behavior for each of them.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala (hunk 10 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:20]: I think you're right, let's keep `SystemMessageData` internal and expose this instead

**Comment by leogrim** [comment]
After chatting with @camhashemi @DerekBurch and @ryanpbrewster to better understand the requirements / benefits, we could expand on `DescriptionElements`, adding element types and richer elements extending `BasicElement`.

File: kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+  case object Firefox extends ActivitySource("Firefox")
+  case object Safari extends ActivitySource("Safari")
+
+  val all = _all
+  def apply(str: String) = all.find(_.value == str).get
+
+  implicit val writes: Writes[ActivitySource] = Writes { o => JsString(o.value) }
+}
+
+sealed abstract class ActivityElement extends DescriptionElement
```
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/ActivityElements.scala:44]: @leogrim @ryanpbrewster updated. I'm going to refactor as we go to re-use whatever can be in `DescriptionElements`
